proposals/Exceptions.md

```diff
@@ -215,39 +216,57 @@ exception on top of the stack is popped and then thrown.
 
 ### Exception data extraction
 
-The `if_except` block begins with an `if_except` instruction, and has two
-instruction blocks,
-
-That is, the forms of an if_except block is:
+The `br_on_exn` instruction is a conditional branch that checks the exception
+tag of an exception on top of the stack, in the form of:
 
 ```
-if_except block_type except_index
-  Instruction*
-end
+br_on_exn label except_index
+```
+
+The `br_on_exn` instruction checks the exception tag of an `except_ref` on top
+of the stack if it matches the given exception index. If it does, it branches
+out to the label referenced by the instruction (In the binary form, the label
+will be converted to a relative depth immediate, like other branch
+instructions), and while doing that, pops exception values from the `except_ref`
+and places them on top of the stack. In order to use these popped values, the
+block signature of the branch target has to match the exception types - because
+it receives the exception arguments as branch operands. If the exception tag
+does not match, the `except_ref` value remains on the stack. For example, when
+an `except_ref` contains an exception of type (i32 i64), the target block
+signature should be (i32 i64) as well, as in the following example:
 
-if_except block_type except_index
-  Instruction*
-else
-  Instruction*
+```
+block $l (result i32 i64)
```

Comment:
Even though #58 suggested the multi-value proposal is a prerequisite for this and this example also uses a multi-value return block, the multi-value proposal is not necessarily a prerequisite for this exception handling proposal; we can do with a single value yielding blocks until the multi value proposal arrives.